taichi/transforms/constant_fold.cpp

```diff
@@ -135,8 +65,85 @@ class ConstantFold : public BasicStmtVisitor {
+      insert_and_erase(stmt, res);                                           \
+    } else if (dt->is_primitive(PrimitiveTypeID::u32) ||                     \
+               dt->is_primitive(PrimitiveTypeID::u64)) {                     \
+      auto res = TypedConstant(                                              \
+          dst_type, PREFIX(lhs->val.val_uint() OP_CPP rhs->val.val_uint())); \
+      insert_and_erase(stmt, res);                                           \
+    }                                                                        \
+    break;                                                                   \
+  }
+
+      HANDLE_REAL_AND_INTEGRAL_BINARY(mul, , *)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(add, , +)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(sub, , -)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(truediv, , /)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(floordiv, std::floor, /)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(div, , /)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(cmp_lt, , <)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(cmp_le, , <=)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(cmp_gt, , >)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(cmp_ge, , >=)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(cmp_eq, , ==)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(cmp_ne, , !=)
+
+      HANDLE_REAL_AND_INTEGRAL_BINARY(max, std::max, COMMA)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(min, std::min, COMMA)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(atan2, std::atan2, COMMA)
+      HANDLE_REAL_AND_INTEGRAL_BINARY(pow, std::pow, COMMA)
+#undef HANDLE_REAL_AND_INTEGRAL_BINARY
+
+#define HANDLE_INTEGRAL_BINARY(OP_TYPE, PREFIX, OP_CPP)                      \
+  case BinaryOpType::OP_TYPE: {                                              \
+    if (dt->is_primitive(PrimitiveTypeID::i32) ||                            \
+        dt->is_primitive(PrimitiveTypeID::i64)) {                            \
+      auto res = TypedConstant(                                              \
+          dst_type, PREFIX(lhs->val.val_int() OP_CPP rhs->val.val_int()));   \
+      insert_and_erase(stmt, res);                                           \
+    } else if (dt->is_primitive(PrimitiveTypeID::u32) ||                     \
+               dt->is_primitive(PrimitiveTypeID::u64)) {                     \
+      auto res = TypedConstant(                                              \
+          dst_type, PREFIX(lhs->val.val_uint() OP_CPP rhs->val.val_uint())); \
+      insert_and_erase(stmt, res);                                           \
+    }                                                                        \
+    break;                                                                   \
+  }
+
+      HANDLE_INTEGRAL_BINARY(mod, , %)
+      HANDLE_INTEGRAL_BINARY(bit_and, , &)
+      HANDLE_INTEGRAL_BINARY(bit_or, , |)
+      HANDLE_INTEGRAL_BINARY(bit_xor, , ^)
+      HANDLE_INTEGRAL_BINARY(bit_shl, , <<)
+      HANDLE_INTEGRAL_BINARY(bit_shr, , >>)
```

Comment:
`>>` in C++ is sar if the data type is signed, not shr

Reply:
`bit_shr` is transformed into `bit_sar` in `demote_operation` but this pass is after `constant_fold` so we have to handle it correctly here. Though I think this logic is a bit strange. Maybe it should be handled in `type_check` like `truediv` so we don't need to handle it here. Also it has a bug (unrelated to this PR) ...... https://github.com/taichi-dev/taichi/issues/7650